src/waveform/renderers/qopengl/waveformrendererfiltered.cpp

```diff
@@ -0,0 +1,255 @@
+#include "waveform/renderers/qopengl/waveformrendererfiltered.h"
+
+#include <iostream>
+
+#include "track/track.h"
+#include "util/math.h"
+#include "waveform/waveform.h"
+#include "waveform/waveformwidgetfactory.h"
+#include "waveform/widgets/qopengl/waveformwidget.h"
+#include "widget/wskincolor.h"
+#include "widget/wwidget.h"
+
+using namespace qopengl;
+
+WaveformRendererFiltered::WaveformRendererFiltered(
+        WaveformWidgetRenderer* waveformWidget)
+        : WaveformRendererSignalBase(waveformWidget) {
+}
+
+WaveformRendererFiltered::~WaveformRendererFiltered() {
+}
+
+void WaveformRendererFiltered::onSetup(const QDomNode& node) {
+    Q_UNUSED(node);
+}
+
+void WaveformRendererFiltered::initializeGL() {
+    QString vertexShaderCode = QStringLiteral(R"--(
+uniform mat4 matrix;
+attribute vec4 position;
+void main()
+{
+    gl_Position = matrix * position;
+}
+)--");
+
+    QString fragmentShaderCode = QStringLiteral(R"--(
+uniform vec4 color;
+void main()
+{
+    gl_FragColor = color;
+}
+)--");
+
+    initShaders(vertexShaderCode, fragmentShaderCode);
+}
+
+inline void WaveformRendererFiltered::addRectangle(
+        QVector<float>& lines,
+        float x1,
+        float y1,
+        float x2,
+        float y2) {
+    lines.push_back(x1);
+    lines.push_back(y1);
+
+    lines.push_back(x2);
+    lines.push_back(y1);
+
+    lines.push_back(x1);
+    lines.push_back(y2);
+
+    lines.push_back(x1);
+    lines.push_back(y2);
+
+    lines.push_back(x2);
+    lines.push_back(y2);
+
+    lines.push_back(x2);
+    lines.push_back(y1);
+}
+
+void WaveformRendererFiltered::renderGL() {
+    TrackPointer pTrack = m_waveformRenderer->getTrackInfo();
+    if (!pTrack) {
+        return;
+    }
+
+    ConstWaveformPointer waveform = pTrack->getWaveform();
+    if (waveform.isNull()) {
+        return;
+    }
+
+    const int dataSize = waveform->getDataSize();
+    if (dataSize <= 1) {
+        return;
+    }
+
+    const WaveformData* data = waveform->data();
+    if (data == nullptr) {
+        return;
+    }
+
+    const float devicePixelRatio = m_waveformRenderer->getDevicePixelRatio();
+    const int length = static_cast<int>(m_waveformRenderer->getLength() * devicePixelRatio);
+
+    // Not multiplying with devicePixelRatio will also work, and on retina
+    // displays 2 pixels will be used to represent each block of samples. This
+    // is also what is done for the beat grid and the markers. const int n =
+    // static_cast<int>(static_cast<float>(m_waveformRenderer->getLength()));
+
+    const double firstVisualIndex = m_waveformRenderer->getFirstDisplayedPosition() * dataSize;
+    const double lastVisualIndex = m_waveformRenderer->getLastDisplayedPosition() * dataSize;
+
+    // Represents the # of waveform data points per horizontal pixel.
+    const double gain = (lastVisualIndex - firstVisualIndex) / static_cast<double>(length);
```

Comment:
I don't understand how this calculation is supposed to work. How does essentially the width of the waveform, correspond to the height?

Reply:
height is not part of this. the word gain (also copy-paste from legacy renderers) is definitely confusing.
how about delta ? or visualIndexIncrementPerPixel